### ISSUE TITLE
flash attention is not work in vggt

### ISSUE THREAD

**Issue by twtynije** [facebookresearch/vggt#156]
Hello, I'm the student who previously inquired about the issue with Flash Attention not working in VGGT. I have installed the Windows version of the Flash Attention wheel file and successfully passed code tests confirming it works properly. However, I'm still encountering errors when running your code. This is quite confusing for me. Could you please provide some suggestions?
**This is the test code.**
`import torch
from flash_attn import flash_attn_func
import time

# 配置参数
batch_size = 2
seq_len = 2048
num_heads = 8
head_dim = 64
dtype = torch.float16
device = "cuda"

# 生成随机输入数据
query = torch.randn(
    batch_size, seq_len, num_heads, head_dim, device=device, dtype=dtype
)
key = torch.randn(batch_size, seq_len, num_heads, head_dim, device=device, dtype=dtype)
value = torch.randn(
    batch_size, seq_len, num_heads, head_dim, device=device, dtype=dtype
)

# Flash Attention计算
flash_output = flash_attn_func(
    query, key, value, dropout_p=0.0, causal=True, softmax_scale=None
)

# 标准注意力实现
def standard_attention(q, k, v, causal=True):
    q = q.transpose(1, 2)
    k = k.transpose(1, 2)
    v = v.transpose(1, 2)

    scores = torch.matmul(q, k.transpose(-1, -2)) / (head_dim**0.5)

    if causal:
        mask = torch.triu(
            torch.ones(seq_len, seq_len, dtype=torch.bool, device=device), diagonal=1
        )
        scores = scores.masked_fill(mask[None, None, :, :], float("-inf"))

    attn_weights = torch.softmax(scores, dim=-1)
    context = torch.matmul(attn_weights, v)
    return context.transpose(1, 2)

# 验证输出一致性
standard_output = standard_attention(query, key, value, causal=True)

print("形状是否一致:", flash_output.shape == standard_output.shape)
print("最大绝对误差:", (flash_output - standard_output).abs().max().item())
print(
    "是否在容差范围内:",
    torch.allclose(flash_output, standard_output, rtol=1e-3, atol=1e-3),
)

# 增强版性能测试函数
def benchmark(fn, *args, repetitions=100, **kwargs):
    # 清理GPU缓存
    torch.cuda.empty_cache()

    # 预热
    for _ in range(10):
        _ = fn(*args, **kwargs)
    torch.cuda.synchronize()

    # 使用CUDA事件精确计时
    start_event = torch.cuda.Event(enable_timing=True)
    end_event = torch.cuda.Event(enable_timing=True)

    start_event.record()
    for _ in range(repetitions):
        _ = fn(*args, **kwargs)
    end_event.record()
    torch.cuda.synchronize()

    # 返回平均时间（毫秒）
    return start_event.elapsed_time(end_event) / repetitions

# 测试Flash Attention性能
flash_time = benchmark(
    flash_attn_func,
    query,
    key,
    value,  # 前三个位置参数
    causal=True,  # 关键字参数
    repetitions=100,
)

# 测试标准注意力性能
standard_time = benchmark(
    standard_attention,
    query,
    key,
    value,  # 前三个位置参数
    causal=True,  # 关键字参数
    repetitions=100,
)
**This is the output result**
print(f"\n性能对比:")
print(f"Flash Attention平均耗时: {flash_time:.3f} ms")
print(f"标准注意力平均耗时: {standard_time:.3f} ms")
print(f"加速比: {standard_time / flash_time:.1f}x")`
This is the output result
形状是否一致: True
最大绝对误差: 0.001953125
是否在容差范围内: True

性能对比:
Flash Attention平均耗时: 0.977 ms
标准注意力平均耗时: 3.995 ms
加速比: 4.1x

**I noticed that your code uses PyTorch 2.3.1, while I'm using PyTorch 2.4.0.
I modified torch.cuda.amp.autocast in the code to torch.amp.autocast('cuda', ...) and made no other changes.**

**cuda=12.4,  flash attention=2.7.4， pytorch=2.4.0, python=3.11.11**

![Image](https://github.com/user-attachments/assets/39166174-0155-4149-bb9c-173b689acb40)

![Image](https://github.com/user-attachments/assets/6e7eebcc-085c-4604-a5b0-047d35c90dd3)
